### PR TITLE
Fix campaign realtime events

### PR DESCRIPTION
**Pull request by Enc3phale**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Automated tests included? | N
| Related user documentation PR URL | N
| Related developer documentation PR URL | N
| Issues addressed (#s or URLs) | N
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
On campaign, realtime events is launched no matter if previous action has been executed and no matter if the taken path is positive or negative.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Create a contact without tag
2. Create a segment with this contact
3. Create a campaign like this
screenshot_2018-11-29 edit campaign mautic

4. Launch command
php app/console m:s:u
php app/console m:c:u
php app/console m:c:t
5. Visit the page that you choose in the campaign event
6. Events `Visits a page` and `Adjust contact points` is triggered

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com/6940)
2. Repeat step above
3. Events are not fired

### CODE REVIEW

File: app/bundles/CampaignBundle/Entity/Event.php [hunk 1]
```diff
@@ -635,6 +635,31 @@ public function getLog()
...
+     * @return LeadEventLog|null
+     */
+    public function getLogByContactAndRotation(Contact $contact, $rotation)
+    {
+        $criteria = Criteria::create()
+            ->where(Criteria::expr()->eq('lead', $contact))
+            ->andWhere(Criteria::expr()->eq('rotation', $rotation))
+            ->orderBy(['id' => 'DESC'])
+            ->setMaxResults(1);
+        ;
```
Comment (heathdutton) [inline app/bundles/CampaignBundle/Entity/Event.php]: Unintentional semicolon?

File: app/bundles/CampaignBundle/Executioner/RealTimeExecutioner.php [hunk 2]
```diff
@@ -240,14 +249,36 @@ private function evaluateDecisionForContact(Event $event, $passthrough = null, $
...
+
+            if ($log === null) {
+                throw new DecisionNotApplicableException("Parent {$parentEvent->getId()} has not been fired, event {$event->getId()} should not be fired.");
+            }
+
+            $pathTaken   = (int)$log->getNonActionPathTaken();
+            $isScheduled = (int)$log->getIsScheduled();
+
+            if ($isScheduled === 1) {
+                throw new DecisionNotApplicableException("Parent {$parentEvent->getId()} has not been fired, event {$event->getId()} should not be fired.");
```
Comment (heathdutton) [inline app/bundles/CampaignBundle/Executioner/RealTimeExecutioner.php]: is_scheduled will be 1 if the event was scheduled and later executed, so you need to also check `$log->getTriggerDate() < now()` or similar. In other words "is_scheduled" doesn't change to a 0 when the event happens.

File: app/bundles/CampaignBundle/Entity/Event.php [hunk 3]
```diff
@@ -635,6 +635,31 @@ public function getLog()
...
+     * @param Contact $contact
+     * @param $rotation
+     * @return LeadEventLog|null
+     */
+    public function getLogByContactAndRotation(Contact $contact, $rotation)
+    {
+        $criteria = Criteria::create()
+            ->where(Criteria::expr()->eq('lead', $contact))
+            ->andWhere(Criteria::expr()->eq('rotation', $rotation))
+            ->orderBy(['id' => 'DESC'])
```
Comment (heathdutton) [inline app/bundles/CampaignBundle/Entity/Event.php]: This orderBy may be superfluous and decrease performance of the query.

File: app/bundles/CampaignBundle/Entity/Event.php (hunk 1 above)
Comment (Enc3phale) [reply app/bundles/CampaignBundle/Entity/Event.php]: Fixed

File: app/bundles/CampaignBundle/Entity/Event.php (hunk 3 above)
Comment (Enc3phale) [reply app/bundles/CampaignBundle/Entity/Event.php]: Fixed

File: app/bundles/CampaignBundle/Executioner/RealTimeExecutioner.php (hunk 2 above)
Comment (Enc3phale) [reply app/bundles/CampaignBundle/Executioner/RealTimeExecutioner.php]: I have made a lot of test and when event is triggered `is_scheduled` column is set to 0.

Comment (Enc3phale) [reply app/bundles/CampaignBundle/Executioner/RealTimeExecutioner.php]: Ex: third and fourth are scheduled events
screenshot_2018-11-29 127 0 0 1 127 0 0 1 mautic-wmk campaign_lead_event_log phpmyadmin 4 7 0

Comment (heathdutton) [reply app/bundles/CampaignBundle/Executioner/RealTimeExecutioner.php]: How is the even triggered though? Real time, or campaign executioner? First attempt or second? There are different kinds of scheduled events. There are events that are intentionally delayed, then there are events that fail (due to integration issue or similar), and *become* scheduled. 

For example, I have over 300 million campaign_lead_event_log entries in production with is_scheduled set to 1 that were triggered in the past. They are not counted as events to be triggered by the campaign executioner.

Unless... something has recently changed drastically in staging, and there's a massive migration coming my way... 😨

Comment (heathdutton) [reply app/bundles/CampaignBundle/Executioner/RealTimeExecutioner.php]: You were right about this... I was thinking of a scenario of a non-scheduled action which can become scheduled when there's a failure... Core doesn't behave the same way as it did last I looked at this.

**Comment by Enc3phale** [comment]
Oh yes, I will fix that

**Pull request merged**